src/mappings/horizonStaking.ts

```diff
@@ -72,3 +83,135 @@ export function handleHorizonStakeWithdrawn(event: HorizonStakeWithdrawn): void
+    indexer.save()
+
+    dataService.totalTokensProvisioned = dataService.totalTokensProvisioned.plus(event.params.tokens)
+    dataService.save()
+
+    graphNetwork.totalTokensProvisioned = graphNetwork.totalTokensProvisioned.plus(event.params.tokens)
+    graphNetwork.save()
+
+    provision.tokensProvisioned = provision.tokensProvisioned.plus(event.params.tokens)
+    provision.save()
+}
+
+export function handleProvisionThawed(event: ProvisionThawed): void {
+    let graphNetwork = createOrLoadGraphNetwork(event.block.number, event.address)
+    let dataService = createOrLoadDataService(event.params.verifier)
+    let indexer = Indexer.load(event.params.serviceProvider.toHexString())!
+    let provision = createOrLoadProvision(event.params.serviceProvider, event.params.verifier, event.block.timestamp)
+
+    indexer.thawingTokens = indexer.thawingTokens.plus(event.params.tokens)
+    indexer.save()
+
+    dataService.totalTokensThawing = dataService.totalTokensThawing.plus(event.params.tokens)
+    dataService.save()
+
+    graphNetwork.totalTokensThawing = graphNetwork.totalTokensThawing.plus(event.params.tokens)
+    graphNetwork.save()
+
+    provision.tokensThawing = provision.tokensThawing.plus(event.params.tokens)
+    provision.save()
+}
+
+export function handleTokensDeprovisioned(event: TokensDeprovisioned): void {
+    let graphNetwork = createOrLoadGraphNetwork(event.block.number, event.address)
+    let dataService = createOrLoadDataService(event.params.verifier)
+    let indexer = Indexer.load(event.params.serviceProvider.toHexString())!
+    let provision = createOrLoadProvision(event.params.serviceProvider, event.params.verifier, event.block.timestamp)
+
+    indexer.provisionedTokens = indexer.provisionedTokens.minus(event.params.tokens)
+    indexer.save()
+
+    dataService.totalTokensProvisioned = dataService.totalTokensProvisioned.minus(event.params.tokens)
+    dataService.save()
+
+    graphNetwork.totalTokensProvisioned = graphNetwork.totalTokensProvisioned.minus(event.params.tokens)
+    graphNetwork.save()
+
+    provision.tokensProvisioned = provision.tokensProvisioned.minus(event.params.tokens)
+    provision.save()
+}
+
+export function handleProvisionParametersSet(event: ProvisionParametersSet): void {
+    // To Do
+}
+
+export function handleProvisionParametersStaged(event: ProvisionParametersStaged): void {
+    // To Do
+}
+
+export function handleProvisionSlashed(event: ProvisionSlashed): void {
+    let graphNetwork = createOrLoadGraphNetwork(event.block.number, event.address)
+    let dataService = createOrLoadDataService(event.params.verifier)
+    let indexer = Indexer.load(event.params.serviceProvider.toHexString())!
+    let provision = createOrLoadProvision(event.params.serviceProvider, event.params.verifier, event.block.timestamp)
+
+    // Due to thawing tokens potentially getting cancelled, we will need to figure the thawing situation
```

Comment:
Would it help if we emitted extra events whenever the thawing cancelling happens? 

We could do `event ThawingPoolReset(string type, uint256 newNonce)`

Reply:
IIRC the main issue was figuring out both if the thaw requests got cancelled (which I seem to recall we discussed it and it was under a very specific condition that all tokens were gone from the slash) but also which thaw requests have to be canceled, since we don't delete thaw request entities, and loading all of them up just to check one by one which one was fulfilled already and which ones were still active doesn't make a lot of sense.